hazelcast/src/main/java/com/hazelcast/jet/impl/MasterContext.java

```diff
@@ -147,8 +149,17 @@ public JobStatus jobStatus() {
         return jobStatus;
     }
 
-    void setJobStatus(JobStatus jobStatus) {
+    void setJobStatus(JobStatus jobStatus, String description, boolean userRequested) {
+        JobStatus oldStatus = this.jobStatus;
         this.jobStatus = jobStatus;
+        jobEventService.publishEvent(jobId, oldStatus, jobStatus, description, userRequested);
+        if (jobStatus.isTerminal()) {
+            jobEventService.removeAllEventListeners(jobId);
```

Comment:
So, we are removing the handler on the server side, but the registration on the client side is there until the user calls to remove the listener, right?

Reply:
I resolved this issue by converting `ClientJobStatusEventHandler` to an interceptor, which removes the client event registration upon detecting a terminal job status event.

Reply:
By the way, I personally find the interceptor solution hacky and actually tried to develop an operation-like solution. However, to my knowledge, we have
1. `Operation`s for inter-member communication,
2. `ClientMessage` + `MessageTask`s for client-to-member communication, and
3. `ClientMessage` + `EventHandler`s for member-to-client communication.

Introducing a new `ClientMessage` holding just the `serverRegistrationId` and registering the corresponding `EventHandler` when initializing the `ClientListenerService` will solve this issue, but create other problems about rolling-upgrade compatibility. It is, I think, apparent that we need some standardized way for member-to-client communication other than event propagation. If there is already a way, please let me know.

Reply:
How would you get the server registration id before the client starts? And, I didn't get how that helps us clean up the registrations that will never fire more events.

Reply:
Normally, users should explicitly deregister event listeners to avoid memory leaks. Some event types and their listeners have the notion of _terminal event_, for which auto-deregistration can be offered to the user. To accomplish the client-side of a deregistration, the coordinator member can send a `ClientMessage`, which holds the server-side registration id, to the corresponding `EventHandler` in the subscriber client. Recall that the member only knows the server-side registration id, whereas the client knows both server- and client-side registration id's.

Reply:
Is this different than an event holding the server registration id? And, why we would need to know the server registration id in that case? Because, we would need that id if we were to send listener deregistration requests from the client to server. But, I assume in that case, we would not need to send anything because we know the terminal event is received and the server-side registration is most probably cleaned, right?

Reply:
> Is this different than an event holding the server registration id?

Yes, of course. Event propagation is performed by the `EventService` in the server-side and `EventHandler`s in the client-side. `EventHandler`s process `ClientMessage`s, which aren't necessarily "events."

> And, why we would need to know the server registration id in that case?

The deregistration would be easier if we know the client registration id in member-to-client communication, but the members only know the server registration id's, so there is no choice.

> Because, we would need that id if we were to send listener deregistration requests from the client to server.

No, the direction of the communication in this issue is opposite: member-to-client.

> But, I assume in that case, we would not need to send anything because we know the terminal event is received and the server-side registration is most probably cleaned, right?

Yes, exactly. We are not trying to initiate the server-side deregistration from the client-side because this is done by the coordinator member. We want the coordinator also initiate the client-side deregistration, but as I explained, required `EventHandler` will create some problems about rolling-upgrade compatibility. The currently implemented interceptor solution is **simple** because where the problem occurs and where the solution is implemented are the same, which is the client, and **efficient** because it doesn't require additional network calls. I just wanted to hear your ideas about the future of member-to-client communication, which is currently limited to event propagation. By the way, during this conversation, I changed my mind, and now I think that _terminal events_ and _event interceptors_ should be accepted as a design pattern to accomplish auto-deregistration wherever needed. I cannot think of a case in which deregistration is required and it is not appropriate to publish this as a terminal event.